---
title: "PRSsum_Simple"
author: "Nuzulul Kurniansyah & Tamar Sofer"
date: "8/20/2023"
output: md_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



## Introduction

This repository provides information regarding the construction of a
multi weighted and un-weighted polygenic risk score (PRSsum).

Suppose that we have multiple PRSs, all using the same $p$ variants\*, and
we sum them with weights. The formula for the $k_{th}$ PRS:

$$\frac{1}{2p}\sum_{i=1}^{p} g_{i}\beta_{ki}$$

Where note that is standardized according to the number of potential alleles (2 times the number of variants). 

For wighted PRSsum, Suppose that we computed the mean and standard deviation of the PRSs, and they are given by $\mu_{k}$,$\sigma_{k}$ for the $K$ PRS. Noting that the number of variants can change between PRSs, we denote by $p_{k}$ the number of variants in the $k_{th}$ PRS. When summing $K$ PRSs with weights, the expression is:

$$wPRSsum=w_{1}\frac{\frac{1}{2p_{1}}\sum_{i=1}^{p} g_{i}\beta_{1i}-\mu_{1}}{\sigma_{1}}+ \dots+w_{1}\frac{\frac{1}{2p_{k}}\sum_{i=1}^{p}g_{k}\beta_{ki}-\mu_{k}}{\sigma_{k}}$$

This can alternatively be written as a single weighted PRS formula after some rearrangement of terms:
$$wPRSsum=g_{i}\left(\frac{w_{1}}{2p_{1}\sigma_{1}} \beta_{1i} + \dots + \frac{w_{k}}{2p_{k}\sigma_{k}} \beta_{ki}\right)-\left( \frac{w_{1}\mu_{1}}{2p_{1}\sigma_{1}}+\dots+\frac{w_{k}\mu_{k}}{2p_{k}\sigma_{k}} \right)$$


For Unweighted PRSsum, we will compute PRSsum without the weight ($w$)

$$PRSsum=\frac{\frac{1}{2p_{1}}\sum_{i=1}^{p}g_{i}\beta_{1i}-\mu_{1}}{\sigma_{1}}+ \dots+\frac{\frac{1}{2p_{k}}\sum_{i=1}^{p}g_{k}\beta_{ki}-\mu_{k}}{\sigma_{k}}$$

This can alternatively be written as a single unweighted PRS formula after some rearrangement of terms:
$$PRSsum=g_{i}\left(\frac{1}{2p_{1}\sigma_{1}} \beta_{1i} + \dots + \frac{1}{2p_{k}\sigma_{k}} \beta_{ki}\right)-\left( \frac{\mu_{1}}{2p_{1}\sigma_{1}}+\dots+\frac{\mu_{k}}{2p_{k}\sigma_{k}} \right)$$


Finnaly, We provide example data set in ./Example directory and code to generate weighted and unweighted PRSsum.


## Example to construct weighted PRSsum 

We have provided an example of how to construct weighted summations of variant weights

## 1. Prepare the variants weight files.

First, we must align alleles across the study and consolidate them into a unified data frame. Below is an illustrative example of how to accomplish this task. we provided ac function a function for allele matching across the study, which can be found in the file "./Code/match_allele.R". NOTE: Please ensure that the column names in your dataset match those used in our provided example to prevent any errors.

```{r,warning=FALSE,message=FALSE}
library(data.table)
library(dplyr)
library(tidyverse)
```


```{r}
source("Code/match_allele.R")

#Open all the files and modify the column names to align with the function's requirements.
#1. Create the list for all variants weight
studies<-c("AFR","EUR","HIS","FINNGEN")
study_list<-list()
for(study in studies){
  variant_weight_file<- paste0("Example/",study,".txt")
  
  variant_weight<-fread(variant_weight_file, data.table = F)
  colnames(variant_weight)<-c("Chromosome", "rsID","Position", "Allele1","Allele2", "BETA")

  study_list[[study]]<- variant_weight
}

#example of variants weight
head(study_list[["AFR"]])
```

Ensure the PRSsum scaling file in correct format

```{r}
### PRSsum Scaling
PRSsum_Scaling<- fread("Example/PRSsum_Scaling.csv", data.table = F)
PRSsum_Scaling
```

Next, we can ensure that all the alleles match across the study by providing reference SNPs. In this example, we used EUR as the reference SNP. Below example of the reference SNP files:

```{r}
ref_snp<- fread("Example/Reference_snp.txt", data.table = F)
head(ref_snp)
```

Next we can match all the variants weight with the reference SNP

```{r}
variants_weight_clean<- match_allele(refrence_snp=ref_snp, 
                                     list_variants_weight=study_list,
                                     match_by_position=FALSE)

head(variants_weight_clean)
```



